### DISCUSSION TITLE
Docker environment variables

### DISCUSSION THREAD

**Discussion by Nico83500** [drakkan/sftpgo#473] [Q&A] [answered]
Hi,
I'm using the SFTPGo Docker image, it seems to work very well.
I just some questions about environment variables : 
- SFTPGO_COMMON__UPLOAD_MODE : what is the default mode ?
- SFTPGO_COMMON__DEFENDER__ENABLED : is it better to use builtin defender or fail2ban ? Or is it just a personal choice ?
- SFTPGO_DATA_PROVIDER__SKIP_NATURAL_KEYS_VALIDATION : I don't understand if I need to enable this. I'm using MariaDB, I use UTF8 characters everyday and I leave default (so false ?). And I don't see any issue ? Or is it for a specific case ?
Thank you !

**Comment by drakkan** [comment, marked as answer]
> Hi,
> I'm using the SFTPGo Docker image, it seems to work very well.
> I just some questions about environment variables :
> 
> * SFTPGO_COMMON__UPLOAD_MODE : what is the default mode ?

0, I'll add this info to the docs.  Thank yoi for noticing. The default `sftpgo.json` configuration file contains all the default values and it can be used as reference if I forgot to document some default. If you want to review the doc and send a PR with this and other fixes it will be much appreciated

> * SFTPGO_COMMON__DEFENDER__ENABLED : is it better to use builtin defender or fail2ban ? Or is it just a personal choice ?

It is a personal choice, obviously I prefer the built-in defender, it should be better integrated, however I didn't received many feedbacks about it, so I'm not sure if people use it and find it useful.

> * SFTPGO_DATA_PROVIDER__SKIP_NATURAL_KEYS_VALIDATION : I don't understand if I need to enable this. I'm using MariaDB, I use UTF8 characters everyday and I leave default (so false ?). And I don't see any issue ? Or is it for a specific case ?
>   Thank you !

If you set to false you cannot use reserved URI chars within usernames and folder names. Try for example to add a user with `òòò` as username, SFTPGo will return a validation error. There is no database issue or other issue within SFTPGo, any UTF-8 chars is supported. Since these value must be passed in rest API URI and web admin URI they could create issues to some old/broken web browser/client library. SFTPGo should handle them with no issues.

**Reply by Nico83500** [reply]
> 0, I'll add this info to the docs. Thank yoi for noticing. The default `sftpgo.json` configuration file contains all the default values and it can be used as reference if I forgot to document some default. If you want to review the doc and send a PR with this and other fixes it will be much appreciated

Thank you ! I didn't know about sftpgo.json so I'll look at it.

> It is a personal choice, obviously I prefer the built-in defender, it should be better integrated, however I didn't received many feedbacks about it, so I'm not sure if people use it and find it useful.

Ok I'll try defender to compare. What are the mini variables needed to use it ? Just enable ? No need to configure anything if I don't want to tune it ?

> If you set to false you cannot use reserved URI chars within usernames and folder names. Try for example to add a user with `òòò` as username, SFTPGo will return a validation error. There is no database issue or other issue within SFTPGo, any UTF-8 chars is supported. Since these value must be passed in rest API URI and web admin URI they could create issues to some old/broken web browser/client library. SFTPGo should handle them with no issues.

Ok I'll try it. So the only disadvantage to enable it is the support of old/broken client ?

Thank you !

**Reply by drakkan** [reply]
> > 0, I'll add this info to the docs. Thank yoi for noticing. The default `sftpgo.json` configuration file contains all the default values and it can be used as reference if I forgot to document some default. If you want to review the doc and send a PR with this and other fixes it will be much appreciated
> 
> Thank you ! I didn't know about sftpgo.json so I'll look at it.
> 
> > It is a personal choice, obviously I prefer the built-in defender, it should be better integrated, however I didn't received many feedbacks about it, so I'm not sure if people use it and find it useful.
> 
> Ok I'll try defender to compare. What are the mini variables needed to use it ? Just enable ? No need to configure anything if I don't want to tune it ?
> 

The default configuration should be ok, you can tune it as you need for your use case. It depends on your needs

> > If you set to false you cannot use reserved URI chars within usernames and folder names. Try for example to add a user with `òòò` as username, SFTPGo will return a validation error. There is no database issue or other issue within SFTPGo, any UTF-8 chars is supported. Since these value must be passed in rest API URI and web admin URI they could create issues to some old/broken web browser/client library. SFTPGo should handle them with no issues.
> 
> Ok I'll try it. So the only disadvantage to enable it is the support of old/broken client ?
> 

Yes, correct

> Thank you !
